src/genesis/mod.rs

```diff
@@ -138,42 +145,71 @@ where
 /// header.
 async fn load_and_retrieve_header<DB, R>(
     store: DB,
-    mut reader: R,
+    reader: R,
     skip_load: bool,
+    chunk_size: ChunkSize,
+    buffer_size: BufferSize,
 ) -> anyhow::Result<(Vec<Cid>, Option<usize>)>
 where
-    DB: Blockstore + Send + Sync + 'static,
+    DB: Blockstore + Send + 'static,
     R: AsyncRead + Send + Unpin,
 {
     let result = if skip_load {
-        (CarReader::new(&mut reader).await?.header.roots, None)
+        (CarReader::new(reader).await?.header.roots, None)
     } else {
-        let (roots, n_records) = forest_load_car(store, &mut reader).await?;
+        let (roots, n_records) = forest_load_car(store, reader, chunk_size, buffer_size).await?;
         (roots, Some(n_records))
     };
 
     Ok(result)
 }
 
-pub async fn forest_load_car<DB, R>(store: DB, reader: R) -> anyhow::Result<(Vec<Cid>, usize)>
+fn car_stream<R: futures::AsyncRead + Send + Unpin>(
+    reader: CarReader<R>,
+) -> impl Stream<Item = anyhow::Result<fvm_ipld_car::Block>> {
+    stream::unfold(reader, |mut reader| async move {
+        reader
+            .next_block()
+            .await
+            .map_err(anyhow::Error::from)
+            .transpose()
+            .map(|result| (result, reader))
+    })
+}
+
+pub async fn forest_load_car<DB, R>(
+    store: DB,
+    reader: R,
+    ChunkSize(chunk_size): ChunkSize,
+    BufferSize(buffer_size): BufferSize,
+) -> anyhow::Result<(Vec<Cid>, usize)>
 where
     R: futures::AsyncRead + Send + Unpin,
-    DB: Blockstore + Send + Sync + 'static,
+    DB: Blockstore + Send + 'static,
 {
-    // 1GB
-    const BUFFER_CAPCITY_BYTES: usize = 1024 * 1024 * 1024;
-
-    let (tx, rx) = flume::bounded(100);
-    let write_task =
-        tokio::spawn(async move { store.buffered_write(rx, BUFFER_CAPCITY_BYTES).await });
     let mut car_reader = CarReader::new(reader).await?;
+    let header = std::mem::take(&mut car_reader.header);
     let mut n_records = 0;
-    while let Some(block) = car_reader.next_block().await? {
-        debug!("Importing block: {}", block.cid);
-        n_records += 1;
-        tx.send_async((block.cid, block.data)).await?;
-    }
-    drop(tx);
-    write_task.await??;
-    Ok((car_reader.header.roots, n_records))
+
+    let sink = futures::sink::unfold(
+        store,
+        |store, blocks: Vec<fvm_ipld_car::Block>| async move {
+            tokio::task::spawn_blocking(move || {
+                store.put_many_keyed(blocks.into_iter().map(|block| (block.cid, block.data)))?;
+                Ok(store)
+            })
+            .await?
+        },
+    );
+
+    // Stream key-value pairs from the CAR file and commit them in chunks. Try
+    // to maintain a buffer of a few chunks to avoid read-stalling.
+    car_stream(car_reader)
+        .inspect(|_| n_records += 1)
+        .chunks(chunk_size)
+        .map(|vec| vec.into_iter().collect::<anyhow::Result<Vec<_>>>())
+        .forward(sink.buffer(buffer_size))
```

Comment:
I don't understand why we have bufferring _and_ chunking - could you add add a doc-comment?

In my head this code is the following, and I don't understand what the introduction of a `sink` does for us
```rust
stream
.chunks(100)
.map(|vec|vec.into_iter().collect())
.try_for_each_concurrent(|vec_of_blocks| tokio::spawn_blocking(||store.put_many_keyed())
```

Reply:
I did try to explain it in the comments: It's to avoid read-stalling. Reading has higher throughput than writing, but there's quite a bit of latency. A few chunks should be prepared while data is being written to the DB to prevent stalling and hide this latency.

Using `try_for_each_concurrent` with no limits will exhaust the pool of blocking threads. Using a limit is functionally the same as a buffer: The `FuturesUnordered` becomes the buffer. However, this approach requires `DB` to be `Sync,` and the db handle has to be stored in an `Arc`.

TL;DR; `try_for_each_concurrent` can solve the same problem but I feel a sink is prettier (fewer constraints, same number of LoC).